plugins/ingest/src/main/java/org/elasticsearch/ingest/IngestDocument.java

```diff
@@ -332,6 +355,16 @@ private void setFieldValue(String path, Object value, boolean append) {
         throw new IllegalArgumentException("field [" + path + "] of type [" + object.getClass().getName() + "] cannot be cast to [" + clazz.getName() + "]");
     }
 
+    private Map<String, Object> createTemplateModel() {
+        Map<String, Object> model = new HashMap<>(sourceAndMetadata);
+        // TODO: I think we should split source from meta data? because now also ES metadata is under the _source key...
+        model.put("_source", sourceAndMetadata);
+        // If there is a field in the source with the name '_ingest' it gets overwritten here,
+        // if access to that field is required then it get accessed via '_source._ingest'
+        model.put("_ingest", ingestMetadata);
+        return model;
```

Comment:
I think the goal here is to have exactly the same behaviour that we have in IngestDocument (FieldPath). The way IngestDocument works is that _id and _source._id are equivalent, so I think everything is fine. the _source prefix won't be used in practice that often, it's there only to make sure that _source._ingest can be accessed, and one cannot really have _id etc. as part of the source.
The only thing that we could improve here is that if we change logic in FieldPath, we have to remember to adapt what is exposed to templates too, which we will most likely forget about :) I don't have a good solution for this yet.